handlers/utils/compute_utils.go

```diff
@@ -0,0 +1,1186 @@
+	if !ok {
+		return types.ImageInspect{}, errors.New("field not exist")
+	}
+	var dockerImage model.DockerImage
+	mapstructure.Decode(imageJSON, &dockerImage)
+	// imageName := fmt.Sprintf("%s:%s", dockerImage.FullName)
+	existing, _, err := client.ImageInspectWithRaw(context.Background(), dockerImage.FullName, false)
+	if err != nil {
+		logrus.Error(err)
+	}
+	if params.Mode == "cached" && err == nil {
+		return existing, nil
+	}
+	if params.Complete {
+		logrus.Info("complete")
+		_, err1 := client.ImageRemove(context.Background(), dockerImage.FullName+params.Tag, types.ImageRemoveOptions{Force: true})
+		if err1 != nil {
+			logrus.Error(err1)
+		}
+		return types.ImageInspect{}, nil
+	}
+	pullImage(&params.Image, progress)
+
+	if len(params.Tag) > 0 {
+		imageInfo := parseRepoTag(dockerImage.FullName)
+		repoTag := fmt.Sprintf("%s:%s", imageInfo["repo"], imageInfo["tag"]+params.Tag)
+		logrus.Info(repoTag)
+		client.ImageTag(context.Background(), dockerImage.FullName, repoTag)
+	}
+
+	inspect, _, err2 := client.ImageInspectWithRaw(context.Background(), dockerImage.FullName, false)
+	logrus.Infof("image inspect %v", inspect)
+	return inspect, err2
+}
+
+func createContainerConfig(imageTag string, command strslice.StrSlice, createConfig map[string]interface{}) *container.Config {
+	createConfig["image"] = imageTag
+	var config container.Config
+	err := mapstructure.Decode(createConfig, &config)
+	res, _ := json.Marshal(config)
+	logrus.Infof("config created %v", string(res))
+	if err != nil {
+		logrus.Error(err)
+	}
+	return &config
+}
+
+func getImageTag(instance *model.Instance) (string, error) {
+	var dockerImage model.DockerImage
+	mapstructure.Decode(instance.Image.Data["dockerImage"], &dockerImage)
+	imageName := dockerImage.FullName
+	if imageName == "" {
+		return "", errors.New("Can not start container with no image")
+	}
+	return imageName, nil
+}
+
+func isTrue(instance *model.Instance, field string) bool {
+	value, ok := GetFieldsIfExist(instance.Data, "fields", field)
+	if ok {
+		return InterfaceToBool(value)
+	}
+	return false
+}
+
+func setupDNSSearch(startConfig map[string]interface{}, instance *model.Instance) {
+	systemCon := instance.SystemContainer
+	if len(systemCon) > 0 {
+		return
+	}
+	// if only rancher search is specified,
+	// prepend search with params read from the system
+	allRancher := true
+	dnsSearch, ok2 := startConfig["dnsSearch"].([]string)
+	if ok2 {
+		logrus.Info("hello")
+		if dnsSearch == nil || len(dnsSearch) == 0 {
+			return
+		}
+		for _, search := range dnsSearch {
+			if strings.HasSuffix(search, "rancher.internal") {
+				continue
+			}
+			allRancher = false
+			break
+		}
+	} else {
+		return
+	}
+
+	if !allRancher {
+		return
+	}
+
+	// read host's resolv.conf
+	file, err := os.Open("/etc/resolv.conf")
+	if err != nil {
+		logrus.Error(err)
+		return
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
```

Comment:
I have no clue what this test is doing and the logic is very confusing (not your fault, its confusing in ptyhon-agent as well) make sure we have tests that cover the various use cases for this